Sources/Sentry/include/SentryTracerMiddleware.h

```diff
@@ -0,0 +1,41 @@
+#import "SentrySpan.h"
+#import "SentryTracer.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * Protocol used to intercept some of `SentryTracer` steps
+ * in order to modify its behavior and distribute concern to other classes.
+ */
+@protocol SentryTracerMiddleware <NSObject>
+
+@optional
+
+/**
+ * Called when the middleware is attached to a SentryTracer
+ */
+- (void)installForTracer:(SentryTracer *)tracer;
+
+/**
+ * Called when the middleware is removed from a SentryTracer
+ */
+- (void)uninstallForTracer:(SentryTracer *)tracer;
+
+/**
+ Return additional spans to be added to the transaction.
+ */
+- (NSArray<id<SentrySpan>> *)createAdditionalSpansForTrace:(SentryTracer *)tracer;
+
+/**
+ Called when the dead line timeout is triggered
+ */
+- (void)tracerDidTimeout:(SentryTracer *)tracer;
+
+/**
+ * Called when the `SentryTracer` is finished.
+ */
+- (void)tracerDidFinish:(SentryTracer *)tracer;
```

Comment:
`m`: Only the `SentryPerformanceTracker` uses this method. The `SentryTimeToDisplayTracker` implements all other methods except this one. Therefore, it makes sense to split this protocol up into multiple ones. I would combine `tracerDidTimeout` and `tracerDidFinish` into something like `TracerLifecycleEvents`.

Reply:
But then we need multiples collections inside of SentryTracer, one for each type of protocol. Or, in my opinion a worst option, one non typed NSArray.

Reply:
We could also move this to `SentryTracerDelegate` as the `SentryPerformanceTracker` is the only delegate.